{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting landlord-ai\n",
      "  Downloading landlord_ai-0.1.32.tar.gz (13 kB)\n",
      "Building wheels for collected packages: landlord-ai\n",
      "  Building wheel for landlord-ai (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for landlord-ai: filename=landlord_ai-0.1.32-py3-none-any.whl size=18048 sha256=05b75324dc486f3d85c84348306b12f57032e89ef9a5cbb28f200c0ef68b642e\n",
      "  Stored in directory: /home/jupyter/.cache/pip/wheels/6c/36/0c/be61a773f6ee6a6673efc46b95fdf97f835e42be8f714f2e19\n",
      "Successfully built landlord-ai\n",
      "Installing collected packages: landlord-ai\n",
      "  Attempting uninstall: landlord-ai\n",
      "    Found existing installation: landlord-ai 0.1.28\n",
      "    Uninstalling landlord-ai-0.1.28:\n",
      "      Successfully uninstalled landlord-ai-0.1.28\n",
      "Successfully installed landlord-ai-0.1.32\n",
      "Collecting keras.preprocessing\n",
      "  Using cached Keras_Preprocessing-1.1.0-py2.py3-none-any.whl (41 kB)\n",
      "Requirement already satisfied: six>=1.9.0 in /opt/conda/lib/python3.7/site-packages (from keras.preprocessing) (1.14.0)\n",
      "Requirement already satisfied: numpy>=1.9.1 in /opt/conda/lib/python3.7/site-packages (from keras.preprocessing) (1.18.2)\n",
      "Installing collected packages: keras.preprocessing\n",
      "Successfully installed keras.preprocessing\n",
      "Requirement already satisfied: tqdm in /opt/conda/lib/python3.7/site-packages (4.43.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install landlord-ai --upgrade\n",
    "!pip install keras.preprocessing --user\n",
    "!pip install tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import keras\n",
    "from keras.utils import Sequence\n",
    "import numpy as np\n",
    "from keras.layers import *\n",
    "from keras.losses import mean_squared_error\n",
    "from keras.callbacks import *\n",
    "\n",
    "import os\n",
    "\n",
    "import pickle\n",
    "import random\n",
    "from tqdm import tqdm\n",
    "\n",
    "from landlordai.game.player import LearningPlayer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PreppedDataGenerator(Sequence):\n",
    "    def __init__(self, path_id, batch_size=1024, timesteps_length=LearningPlayer.TIMESTEPS):\n",
    "        self.path_id = path_id\n",
    "        self.batch_size = batch_size\n",
    "        self.timesteps_length = timesteps_length\n",
    "        \n",
    "        self.load_cache()\n",
    "        self.on_epoch_end()\n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.cache[0]) // self.batch_size\n",
    "    \n",
    "    def on_epoch_end(self):\n",
    "        p = np.random.permutation(len(self.cache[0]))\n",
    "        \n",
    "        all_history_matrices = self.cache[0][p]\n",
    "        all_move_vectors = self.cache[1][p]\n",
    "        all_hand_vectors = self.cache[2][p]\n",
    "        all_y = self.cache[3][p]\n",
    "        \n",
    "        self.cache = (all_history_matrices, all_move_vectors, all_hand_vectors, all_y)\n",
    "        \n",
    "    def load_cache(self):\n",
    "        with open(self.path_id, 'rb') as f:\n",
    "            history_matrices, move_vectors, hand_vectors, y = pickle.load(f)\n",
    "            \n",
    "        self.cache = (history_matrices, move_vectors, hand_vectors, np.array(y)) \n",
    "        \n",
    "        \n",
    "    def __getitem__(self, index):\n",
    "        \"\"\"Generate one batch of data\n",
    "        :param index: index of the batch\n",
    "        :return: X and y when fitting. X only when predicting\n",
    "        \"\"\"\n",
    "        \n",
    "        history_matrices = self.cache[0][index * self.batch_size: (index + 1) * self.batch_size]\n",
    "        move_vectors = self.cache[1][index * self.batch_size: (index + 1) * self.batch_size]\n",
    "        hand_vectors = self.cache[2][index * self.batch_size: (index + 1) * self.batch_size]\n",
    "        y = self.cache[3][index * self.batch_size: (index + 1) * self.batch_size]\n",
    "        \n",
    "        #return [self.densify(history_matrices), move_vectors, hand_vectors], y\n",
    "        return [history_matrices, move_vectors, hand_vectors], y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_folder = '4_10_actualq1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying gs://landlord_ai/4_10_actualq1_agg/test.pkl...\n",
      "Copying gs://landlord_ai/4_10_actualq1_agg/train.pkl...\n",
      "\\ [2/2 files][ 15.5 GiB/ 15.5 GiB] 100% Done 125.6 MiB/s ETA 00:00:00           \n",
      "Operation completed over 2 objects/15.5 GiB.                                     \n"
     ]
    }
   ],
   "source": [
    "assert data_folder is not None\n",
    "!rm -r ../data/{data_folder}_agg\n",
    "!gsutil -m cp -r gs://landlord_ai/{data_folder}_agg/ ../data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_gen = PreppedDataGenerator('../data/' + data_folder + '_agg/train.pkl', batch_size=1 << 11)\n",
    "test_gen = PreppedDataGenerator('../data/' + data_folder + '_agg/test.pkl', batch_size=1 << 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert not np.allclose(train_gen[1][0][0], train_gen[0][0][0])\n",
    "assert len(train_gen[0][0]) == 3\n",
    "for i in range(3):\n",
    "    get_set = train_gen[0][0][0]\n",
    "    if len(get_set.shape) != 3:\n",
    "        print(get_set)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_model_bidi():\n",
    "    K.clear_session()\n",
    "    GRU_DIM = 128\n",
    "\n",
    "    history_inp = Input((None, LearningPlayer.TIMESTEP_FEATURES), name='history_inp')\n",
    "    move_inp = Input((LearningPlayer.TIMESTEP_FEATURES, ), name='move_inp')\n",
    "    hand_inp = Input((LearningPlayer.HAND_FEATURES, ), name='hand_inp')\n",
    "    gru = Bidirectional(GRU(GRU_DIM, name='gru'), name='bidi')(history_inp)\n",
    "\n",
    "    concat = Concatenate()([gru, move_inp, hand_inp])\n",
    "    hidden1 = Dense(256, activation='relu', name='hidden1')(concat)\n",
    "    hidden2 = Dense(128, activation='relu', name='hidden2')(BatchNormalization(name='bn1')(hidden1))\n",
    "    hidden3 = Dense(64, activation='relu', name='hidden3')(BatchNormalization(name='bn2')(hidden2))\n",
    "\n",
    "    output = Dense(1, activation='linear', name='output')(BatchNormalization(name='bn3')(hidden3))\n",
    "    combined_net = keras.models.Model(inputs=[history_inp, move_inp, hand_inp], outputs=output)\n",
    "    combined_net.compile(loss=keras.losses.mean_squared_error, optimizer='adam', metrics=['mean_squared_error'])\n",
    "    return combined_net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "def has_layer(model, layer):\n",
    "    try:\n",
    "        model.get_layer(layer)\n",
    "        return True\n",
    "    except:\n",
    "        return False\n",
    "\n",
    "def split_model_triage(composite, model_folder):\n",
    "    best_model = keras.models.load_model(composite)\n",
    "    \n",
    "    split_model(best_model, model_folder)\n",
    "    \n",
    "def split_model(best_model, model_folder):\n",
    "    bn1 = best_model.get_layer('bn1')\n",
    "    bn2 = best_model.get_layer('bn2')\n",
    "    bn3 = best_model.get_layer('bn3')\n",
    "    history_net = keras.models.Model(inputs=[best_model.get_layer('history_inp').input], outputs=[best_model.get_layer('bidi').output])\n",
    "\n",
    "    vector_history_inp = Input((best_model.get_layer('bidi').output.shape[1], ), name='vector_history_inp')\n",
    "    \n",
    "    concat = Concatenate()([vector_history_inp, best_model.get_layer('move_inp').output, best_model.get_layer('hand_inp').output])\n",
    "    hidden1 = best_model.get_layer('hidden1')(concat)\n",
    "    hidden2 = best_model.get_layer('hidden2')(bn1(hidden1))\n",
    "    hidden3 = best_model.get_layer('hidden3')(bn2(hidden2))\n",
    "    output = best_model.get_layer('output')(bn3(hidden3))\n",
    "\n",
    "    move_inp = best_model.get_layer('move_inp').input\n",
    "    hand_inp = best_model.get_layer('hand_inp').input\n",
    "    position_net = keras.models.Model(inputs=[vector_history_inp, move_inp, hand_inp], outputs=[output])\n",
    "\n",
    "    history_net.save(str(model_folder / 'history.h5'))\n",
    "    position_net.save(str(model_folder / 'position.h5'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sanity_check_model(combined_file, net_dir):\n",
    "    sanity_set = train_gen[0]\n",
    "    historical_features, move_vectors, hand_vectors = sanity_set[0]\n",
    "    targets = sanity_set[1]\n",
    "\n",
    "    player = LearningPlayer(name='sanity', net_dir=str(net_dir))\n",
    "    \n",
    "    historical_matrix = player.history_net.predict(historical_features, batch_size=1024)\n",
    "\n",
    "    from sklearn import metrics\n",
    "    \n",
    "    error_1 = metrics.mean_squared_error(targets, player.get_position_predictions(historical_matrix, move_vectors, hand_vectors))\n",
    "    \n",
    "    composite = keras.models.load_model(combined_file)\n",
    "    error_2 = metrics.mean_squared_error(targets, composite.predict([historical_features, move_vectors, hand_vectors], batch_size=1024))\n",
    "    print(combined_file, error_1, error_2)\n",
    "    assert np.abs(error_1 - error_2) < 1E-2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "def delete_dir(path):\n",
    "    if not os.path.exists(path):\n",
    "        return\n",
    "    for file in path.iterdir():\n",
    "        os.remove(file)\n",
    "    path.rmdir()\n",
    "    \n",
    "def publish_model(i):\n",
    "    combined_file = data_folder + '_combined_' + str(i) + '.h5'\n",
    "    if os.path.exists(combined_file):\n",
    "        model_folder_name = data_folder + '_model' + str(i)\n",
    "\n",
    "        model_folder_path = Path('../models/', model_folder_name)\n",
    "        delete_dir(model_folder_path)\n",
    "        model_folder_path.mkdir()\n",
    "\n",
    "        split_model_triage(combined_file, model_folder_path)\n",
    "        sanity_check_model(combined_file, model_folder_path)\n",
    "        print(model_folder_name)\n",
    "        subprocess.check_output(['gsutil', 'cp', '-r', '../models/' + model_folder_name + '/*', 'gs://landlord_ai/models/' + model_folder_name])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/50\n",
      "3000/3000 [==============================] - 491s 164ms/step - loss: 0.9706 - mean_squared_error: 0.9706 - val_loss: 0.5127 - val_mean_squared_error: 0.4902\n",
      "\n",
      "Epoch 00001: val_mean_squared_error improved from inf to 0.49016, saving model to 4_10_actualq1_combined_14.h5\n",
      "Epoch 2/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.3283 - mean_squared_error: 0.3283 - val_loss: 0.2712 - val_mean_squared_error: 0.3634\n",
      "\n",
      "Epoch 00002: val_mean_squared_error improved from 0.49016 to 0.36341, saving model to 4_10_actualq1_combined_14.h5\n",
      "Epoch 3/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.2667 - mean_squared_error: 0.2667 - val_loss: 0.2459 - val_mean_squared_error: 0.3346\n",
      "\n",
      "Epoch 00003: val_mean_squared_error improved from 0.36341 to 0.33459, saving model to 4_10_actualq1_combined_14.h5\n",
      "Epoch 4/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.2419 - mean_squared_error: 0.2419 - val_loss: 0.2053 - val_mean_squared_error: 0.2438\n",
      "\n",
      "Epoch 00004: val_mean_squared_error improved from 0.33459 to 0.24382, saving model to 4_10_actualq1_combined_14.h5\n",
      "Epoch 5/50\n",
      "3000/3000 [==============================] - 497s 166ms/step - loss: 0.2269 - mean_squared_error: 0.2269 - val_loss: 0.2433 - val_mean_squared_error: 0.2726\n",
      "\n",
      "Epoch 00005: val_mean_squared_error did not improve from 0.24382\n",
      "Epoch 6/50\n",
      "3000/3000 [==============================] - 488s 163ms/step - loss: 0.2220 - mean_squared_error: 0.2220 - val_loss: 0.3473 - val_mean_squared_error: 0.3123\n",
      "\n",
      "Epoch 00006: val_mean_squared_error did not improve from 0.24382\n",
      "Epoch 7/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.2090 - mean_squared_error: 0.2090 - val_loss: 0.2051 - val_mean_squared_error: 0.2156\n",
      "\n",
      "Epoch 00007: val_mean_squared_error improved from 0.24382 to 0.21559, saving model to 4_10_actualq1_combined_14.h5\n",
      "Epoch 8/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.2072 - mean_squared_error: 0.2072 - val_loss: 0.2418 - val_mean_squared_error: 0.2186\n",
      "\n",
      "Epoch 00008: val_mean_squared_error did not improve from 0.21559\n",
      "Epoch 9/50\n",
      "3000/3000 [==============================] - 494s 165ms/step - loss: 0.1993 - mean_squared_error: 0.1993 - val_loss: 0.2674 - val_mean_squared_error: 0.2224\n",
      "\n",
      "Epoch 00009: val_mean_squared_error did not improve from 0.21559\n",
      "Epoch 10/50\n",
      "3000/3000 [==============================] - 493s 164ms/step - loss: 0.1976 - mean_squared_error: 0.1976 - val_loss: 0.1881 - val_mean_squared_error: 0.2101\n",
      "\n",
      "Epoch 00010: val_mean_squared_error improved from 0.21559 to 0.21011, saving model to 4_10_actualq1_combined_14.h5\n",
      "Epoch 11/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.1929 - mean_squared_error: 0.1929 - val_loss: 0.2018 - val_mean_squared_error: 0.2021\n",
      "\n",
      "Epoch 00011: val_mean_squared_error improved from 0.21011 to 0.20210, saving model to 4_10_actualq1_combined_14.h5\n",
      "Epoch 12/50\n",
      "3000/3000 [==============================] - 493s 164ms/step - loss: 0.1900 - mean_squared_error: 0.1900 - val_loss: 0.1977 - val_mean_squared_error: 0.1954\n",
      "\n",
      "Epoch 00012: val_mean_squared_error improved from 0.20210 to 0.19542, saving model to 4_10_actualq1_combined_14.h5\n",
      "Epoch 13/50\n",
      "3000/3000 [==============================] - 493s 164ms/step - loss: 0.1887 - mean_squared_error: 0.1887 - val_loss: 0.2399 - val_mean_squared_error: 0.1988\n",
      "\n",
      "Epoch 00013: val_mean_squared_error did not improve from 0.19542\n",
      "Epoch 14/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.1829 - mean_squared_error: 0.1829 - val_loss: 0.1821 - val_mean_squared_error: 0.2063\n",
      "\n",
      "Epoch 00014: val_mean_squared_error did not improve from 0.19542\n",
      "Epoch 15/50\n",
      "3000/3000 [==============================] - 493s 164ms/step - loss: 0.1826 - mean_squared_error: 0.1826 - val_loss: 0.3040 - val_mean_squared_error: 0.2136\n",
      "\n",
      "Epoch 00015: val_mean_squared_error did not improve from 0.19542\n",
      "Epoch 00015: early stopping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/keras/engine/saving.py:341: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.\n",
      "  warnings.warn('No training configuration found in save file: '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4_10_actualq1_combined_14.h5 0.18332753 0.18332753\n",
      "4_10_actualq1_model14\n",
      "Epoch 1/50\n",
      "3000/3000 [==============================] - 488s 163ms/step - loss: 0.8727 - mean_squared_error: 0.8727 - val_loss: 0.3879 - val_mean_squared_error: 0.3279\n",
      "\n",
      "Epoch 00001: val_mean_squared_error improved from inf to 0.32790, saving model to 4_10_actualq1_combined_15.h5\n",
      "Epoch 2/50\n",
      "3000/3000 [==============================] - 493s 164ms/step - loss: 0.3069 - mean_squared_error: 0.3069 - val_loss: 0.3525 - val_mean_squared_error: 0.3709\n",
      "\n",
      "Epoch 00002: val_mean_squared_error did not improve from 0.32790\n",
      "Epoch 3/50\n",
      "3000/3000 [==============================] - 488s 163ms/step - loss: 0.2599 - mean_squared_error: 0.2599 - val_loss: 0.3355 - val_mean_squared_error: 0.3100\n",
      "\n",
      "Epoch 00003: val_mean_squared_error improved from 0.32790 to 0.30996, saving model to 4_10_actualq1_combined_15.h5\n",
      "Epoch 4/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.2385 - mean_squared_error: 0.2385 - val_loss: 0.2076 - val_mean_squared_error: 0.2355\n",
      "\n",
      "Epoch 00004: val_mean_squared_error improved from 0.30996 to 0.23546, saving model to 4_10_actualq1_combined_15.h5\n",
      "Epoch 5/50\n",
      "3000/3000 [==============================] - 489s 163ms/step - loss: 0.2235 - mean_squared_error: 0.2235 - val_loss: 0.1960 - val_mean_squared_error: 0.2182\n",
      "\n",
      "Epoch 00005: val_mean_squared_error improved from 0.23546 to 0.21824, saving model to 4_10_actualq1_combined_15.h5\n",
      "Epoch 6/50\n",
      "3000/3000 [==============================] - 481s 160ms/step - loss: 0.2143 - mean_squared_error: 0.2143 - val_loss: 0.2439 - val_mean_squared_error: 0.2099\n",
      "\n",
      "Epoch 00006: val_mean_squared_error improved from 0.21824 to 0.20995, saving model to 4_10_actualq1_combined_15.h5\n",
      "Epoch 7/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.2027 - mean_squared_error: 0.2027 - val_loss: 0.2349 - val_mean_squared_error: 0.2518\n",
      "\n",
      "Epoch 00007: val_mean_squared_error did not improve from 0.20995\n",
      "Epoch 8/50\n",
      "3000/3000 [==============================] - 488s 163ms/step - loss: 0.2049 - mean_squared_error: 0.2049 - val_loss: 0.2294 - val_mean_squared_error: 0.2129\n",
      "\n",
      "Epoch 00008: val_mean_squared_error did not improve from 0.20995\n",
      "Epoch 9/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.1929 - mean_squared_error: 0.1929 - val_loss: 0.1662 - val_mean_squared_error: 0.2132\n",
      "\n",
      "Epoch 00009: val_mean_squared_error did not improve from 0.20995\n",
      "Epoch 00009: early stopping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/keras/engine/saving.py:341: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.\n",
      "  warnings.warn('No training configuration found in save file: '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4_10_actualq1_combined_15.h5 0.20493196 0.20493197\n",
      "4_10_actualq1_model15\n",
      "Epoch 1/50\n",
      "3000/3000 [==============================] - 490s 163ms/step - loss: 0.9123 - mean_squared_error: 0.9123 - val_loss: 0.3892 - val_mean_squared_error: 0.4696\n",
      "\n",
      "Epoch 00001: val_mean_squared_error improved from inf to 0.46965, saving model to 4_10_actualq1_combined_16.h5\n",
      "Epoch 2/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.3019 - mean_squared_error: 0.3019 - val_loss: 0.3791 - val_mean_squared_error: 0.3882\n",
      "\n",
      "Epoch 00002: val_mean_squared_error improved from 0.46965 to 0.38821, saving model to 4_10_actualq1_combined_16.h5\n",
      "Epoch 3/50\n",
      "3000/3000 [==============================] - 489s 163ms/step - loss: 0.3244 - mean_squared_error: 0.3244 - val_loss: 0.2833 - val_mean_squared_error: 0.2909\n",
      "\n",
      "Epoch 00003: val_mean_squared_error improved from 0.38821 to 0.29089, saving model to 4_10_actualq1_combined_16.h5\n",
      "Epoch 4/50\n",
      "3000/3000 [==============================] - 497s 166ms/step - loss: 0.2518 - mean_squared_error: 0.2518 - val_loss: 0.3472 - val_mean_squared_error: 0.2702\n",
      "\n",
      "Epoch 00004: val_mean_squared_error improved from 0.29089 to 0.27024, saving model to 4_10_actualq1_combined_16.h5\n",
      "Epoch 5/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.2317 - mean_squared_error: 0.2317 - val_loss: 0.2103 - val_mean_squared_error: 0.2368\n",
      "\n",
      "Epoch 00005: val_mean_squared_error improved from 0.27024 to 0.23680, saving model to 4_10_actualq1_combined_16.h5\n",
      "Epoch 6/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.2170 - mean_squared_error: 0.2170 - val_loss: 0.2625 - val_mean_squared_error: 0.2437\n",
      "\n",
      "Epoch 00006: val_mean_squared_error did not improve from 0.23680\n",
      "Epoch 7/50\n",
      "3000/3000 [==============================] - 494s 165ms/step - loss: 0.2135 - mean_squared_error: 0.2135 - val_loss: 0.2584 - val_mean_squared_error: 0.2729\n",
      "\n",
      "Epoch 00007: val_mean_squared_error did not improve from 0.23680\n",
      "Epoch 8/50\n",
      "3000/3000 [==============================] - 486s 162ms/step - loss: 0.2028 - mean_squared_error: 0.2028 - val_loss: 0.1700 - val_mean_squared_error: 0.2018\n",
      "\n",
      "Epoch 00008: val_mean_squared_error improved from 0.23680 to 0.20180, saving model to 4_10_actualq1_combined_16.h5\n",
      "Epoch 9/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.2013 - mean_squared_error: 0.2013 - val_loss: 0.1826 - val_mean_squared_error: 0.2307\n",
      "\n",
      "Epoch 00009: val_mean_squared_error did not improve from 0.20180\n",
      "Epoch 10/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1924 - mean_squared_error: 0.1925 - val_loss: 0.1655 - val_mean_squared_error: 0.2129\n",
      "\n",
      "Epoch 00010: val_mean_squared_error did not improve from 0.20180\n",
      "Epoch 11/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.1901 - mean_squared_error: 0.1901 - val_loss: 0.1640 - val_mean_squared_error: 0.1920\n",
      "\n",
      "Epoch 00011: val_mean_squared_error improved from 0.20180 to 0.19197, saving model to 4_10_actualq1_combined_16.h5\n",
      "Epoch 12/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1849 - mean_squared_error: 0.1849 - val_loss: 0.2169 - val_mean_squared_error: 0.1806\n",
      "\n",
      "Epoch 00012: val_mean_squared_error improved from 0.19197 to 0.18063, saving model to 4_10_actualq1_combined_16.h5\n",
      "Epoch 13/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1884 - mean_squared_error: 0.1884 - val_loss: 0.3532 - val_mean_squared_error: 0.2615\n",
      "\n",
      "Epoch 00013: val_mean_squared_error did not improve from 0.18063\n",
      "Epoch 14/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.1785 - mean_squared_error: 0.1785 - val_loss: 0.2047 - val_mean_squared_error: 0.2038\n",
      "\n",
      "Epoch 00014: val_mean_squared_error did not improve from 0.18063\n",
      "Epoch 15/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.1796 - mean_squared_error: 0.1796 - val_loss: 0.1810 - val_mean_squared_error: 0.1856\n",
      "\n",
      "Epoch 00015: val_mean_squared_error did not improve from 0.18063\n",
      "Epoch 00015: early stopping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/keras/engine/saving.py:341: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.\n",
      "  warnings.warn('No training configuration found in save file: '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4_10_actualq1_combined_16.h5 0.15169781 0.15169778\n",
      "4_10_actualq1_model16\n",
      "Epoch 1/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.9265 - mean_squared_error: 0.9265 - val_loss: 0.3519 - val_mean_squared_error: 0.4009\n",
      "\n",
      "Epoch 00001: val_mean_squared_error improved from inf to 0.40090, saving model to 4_10_actualq1_combined_17.h5\n",
      "Epoch 2/50\n",
      "3000/3000 [==============================] - 492s 164ms/step - loss: 0.3020 - mean_squared_error: 0.3020 - val_loss: 0.3013 - val_mean_squared_error: 0.3309\n",
      "\n",
      "Epoch 00002: val_mean_squared_error improved from 0.40090 to 0.33087, saving model to 4_10_actualq1_combined_17.h5\n",
      "Epoch 3/50\n",
      "3000/3000 [==============================] - 485s 162ms/step - loss: 0.2610 - mean_squared_error: 0.2610 - val_loss: 0.2404 - val_mean_squared_error: 0.2662\n",
      "\n",
      "Epoch 00003: val_mean_squared_error improved from 0.33087 to 0.26618, saving model to 4_10_actualq1_combined_17.h5\n",
      "Epoch 4/50\n",
      "3000/3000 [==============================] - 493s 164ms/step - loss: 0.2355 - mean_squared_error: 0.2355 - val_loss: 0.3556 - val_mean_squared_error: 0.2413\n",
      "\n",
      "Epoch 00004: val_mean_squared_error improved from 0.26618 to 0.24130, saving model to 4_10_actualq1_combined_17.h5\n",
      "Epoch 5/50\n",
      "3000/3000 [==============================] - 493s 164ms/step - loss: 0.2210 - mean_squared_error: 0.2210 - val_loss: 0.2638 - val_mean_squared_error: 0.2459\n",
      "\n",
      "Epoch 00005: val_mean_squared_error did not improve from 0.24130\n",
      "Epoch 6/50\n",
      "3000/3000 [==============================] - 485s 162ms/step - loss: 0.2132 - mean_squared_error: 0.2132 - val_loss: 0.1779 - val_mean_squared_error: 0.2090\n",
      "\n",
      "Epoch 00006: val_mean_squared_error improved from 0.24130 to 0.20900, saving model to 4_10_actualq1_combined_17.h5\n",
      "Epoch 7/50\n",
      "3000/3000 [==============================] - 491s 164ms/step - loss: 0.2067 - mean_squared_error: 0.2067 - val_loss: 0.3435 - val_mean_squared_error: 0.2294\n",
      "\n",
      "Epoch 00007: val_mean_squared_error did not improve from 0.20900\n",
      "Epoch 8/50\n",
      "3000/3000 [==============================] - 483s 161ms/step - loss: 0.1984 - mean_squared_error: 0.1984 - val_loss: 0.2403 - val_mean_squared_error: 0.2100\n",
      "\n",
      "Epoch 00008: val_mean_squared_error did not improve from 0.20900\n",
      "Epoch 9/50\n",
      "3000/3000 [==============================] - 492s 164ms/step - loss: 0.1949 - mean_squared_error: 0.1949 - val_loss: 0.2305 - val_mean_squared_error: 0.1964\n",
      "\n",
      "Epoch 00009: val_mean_squared_error improved from 0.20900 to 0.19636, saving model to 4_10_actualq1_combined_17.h5\n",
      "Epoch 10/50\n",
      "3000/3000 [==============================] - 493s 164ms/step - loss: 0.1898 - mean_squared_error: 0.1898 - val_loss: 0.1704 - val_mean_squared_error: 0.1927\n",
      "\n",
      "Epoch 00010: val_mean_squared_error improved from 0.19636 to 0.19270, saving model to 4_10_actualq1_combined_17.h5\n",
      "Epoch 11/50\n",
      "3000/3000 [==============================] - 485s 162ms/step - loss: 0.1879 - mean_squared_error: 0.1879 - val_loss: 0.2192 - val_mean_squared_error: 0.1862\n",
      "\n",
      "Epoch 00011: val_mean_squared_error improved from 0.19270 to 0.18624, saving model to 4_10_actualq1_combined_17.h5\n",
      "Epoch 12/50\n",
      "3000/3000 [==============================] - 492s 164ms/step - loss: 0.1821 - mean_squared_error: 0.1821 - val_loss: 0.1804 - val_mean_squared_error: 0.2021\n",
      "\n",
      "Epoch 00012: val_mean_squared_error did not improve from 0.18624\n",
      "Epoch 13/50\n",
      "3000/3000 [==============================] - 492s 164ms/step - loss: 0.1809 - mean_squared_error: 0.1809 - val_loss: 0.1840 - val_mean_squared_error: 0.2140\n",
      "\n",
      "Epoch 00013: val_mean_squared_error did not improve from 0.18624\n",
      "Epoch 14/50\n",
      "3000/3000 [==============================] - 485s 162ms/step - loss: 0.1772 - mean_squared_error: 0.1772 - val_loss: 0.2399 - val_mean_squared_error: 0.2052\n",
      "\n",
      "Epoch 00014: val_mean_squared_error did not improve from 0.18624\n",
      "Epoch 00014: early stopping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/keras/engine/saving.py:341: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.\n",
      "  warnings.warn('No training configuration found in save file: '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4_10_actualq1_combined_17.h5 0.24334693 0.24334693\n",
      "4_10_actualq1_model17\n",
      "Epoch 1/50\n",
      "3000/3000 [==============================] - 490s 163ms/step - loss: 0.9202 - mean_squared_error: 0.9202 - val_loss: 0.4724 - val_mean_squared_error: 0.4121\n",
      "\n",
      "Epoch 00001: val_mean_squared_error improved from inf to 0.41212, saving model to 4_10_actualq1_combined_18.h5\n",
      "Epoch 2/50\n",
      "3000/3000 [==============================] - 494s 165ms/step - loss: 0.3064 - mean_squared_error: 0.3064 - val_loss: 0.3063 - val_mean_squared_error: 0.2753\n",
      "\n",
      "Epoch 00002: val_mean_squared_error improved from 0.41212 to 0.27534, saving model to 4_10_actualq1_combined_18.h5\n",
      "Epoch 3/50\n",
      "3000/3000 [==============================] - 485s 162ms/step - loss: 0.2624 - mean_squared_error: 0.2624 - val_loss: 0.4779 - val_mean_squared_error: 0.3888\n",
      "\n",
      "Epoch 00003: val_mean_squared_error did not improve from 0.27534\n",
      "Epoch 4/50\n",
      "3000/3000 [==============================] - 494s 165ms/step - loss: 0.2368 - mean_squared_error: 0.2368 - val_loss: 0.2943 - val_mean_squared_error: 0.3217\n",
      "\n",
      "Epoch 00004: val_mean_squared_error did not improve from 0.27534\n",
      "Epoch 5/50\n",
      "3000/3000 [==============================] - 489s 163ms/step - loss: 0.2261 - mean_squared_error: 0.2261 - val_loss: 0.3627 - val_mean_squared_error: 0.3304\n",
      "\n",
      "Epoch 00005: val_mean_squared_error did not improve from 0.27534\n",
      "Epoch 00005: early stopping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/keras/engine/saving.py:341: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.\n",
      "  warnings.warn('No training configuration found in save file: '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4_10_actualq1_combined_18.h5 0.2515535 0.25155348\n",
      "4_10_actualq1_model18\n",
      "Epoch 1/50\n",
      "3000/3000 [==============================] - 489s 163ms/step - loss: 0.8057 - mean_squared_error: 0.8057 - val_loss: 0.2819 - val_mean_squared_error: 0.3803\n",
      "\n",
      "Epoch 00001: val_mean_squared_error improved from inf to 0.38027, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 2/50\n",
      "3000/3000 [==============================] - 492s 164ms/step - loss: 0.3062 - mean_squared_error: 0.3062 - val_loss: 0.3541 - val_mean_squared_error: 0.3440\n",
      "\n",
      "Epoch 00002: val_mean_squared_error improved from 0.38027 to 0.34401, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 3/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.2604 - mean_squared_error: 0.2604 - val_loss: 0.2161 - val_mean_squared_error: 0.2754\n",
      "\n",
      "Epoch 00003: val_mean_squared_error improved from 0.34401 to 0.27544, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 4/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.2363 - mean_squared_error: 0.2363 - val_loss: 0.2790 - val_mean_squared_error: 0.2384\n",
      "\n",
      "Epoch 00004: val_mean_squared_error improved from 0.27544 to 0.23840, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 5/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.2271 - mean_squared_error: 0.2271 - val_loss: 0.2663 - val_mean_squared_error: 0.2786\n",
      "\n",
      "Epoch 00005: val_mean_squared_error did not improve from 0.23840\n",
      "Epoch 6/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.2182 - mean_squared_error: 0.2182 - val_loss: 0.2416 - val_mean_squared_error: 0.2788\n",
      "\n",
      "Epoch 00006: val_mean_squared_error did not improve from 0.23840\n",
      "Epoch 7/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.2073 - mean_squared_error: 0.2073 - val_loss: 0.2221 - val_mean_squared_error: 0.2171\n",
      "\n",
      "Epoch 00007: val_mean_squared_error improved from 0.23840 to 0.21707, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 8/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.2018 - mean_squared_error: 0.2018 - val_loss: 0.2538 - val_mean_squared_error: 0.2557\n",
      "\n",
      "Epoch 00008: val_mean_squared_error did not improve from 0.21707\n",
      "Epoch 9/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1970 - mean_squared_error: 0.1970 - val_loss: 0.2320 - val_mean_squared_error: 0.2160\n",
      "\n",
      "Epoch 00009: val_mean_squared_error improved from 0.21707 to 0.21599, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 10/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1933 - mean_squared_error: 0.1933 - val_loss: 0.1600 - val_mean_squared_error: 0.2156\n",
      "\n",
      "Epoch 00010: val_mean_squared_error improved from 0.21599 to 0.21556, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 11/50\n",
      "3000/3000 [==============================] - 488s 163ms/step - loss: 0.1894 - mean_squared_error: 0.1894 - val_loss: 0.1989 - val_mean_squared_error: 0.1898\n",
      "\n",
      "Epoch 00011: val_mean_squared_error improved from 0.21556 to 0.18980, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 12/50\n",
      "3000/3000 [==============================] - 494s 165ms/step - loss: 0.1832 - mean_squared_error: 0.1832 - val_loss: 0.2009 - val_mean_squared_error: 0.1943\n",
      "\n",
      "Epoch 00012: val_mean_squared_error did not improve from 0.18980\n",
      "Epoch 13/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.1860 - mean_squared_error: 0.1860 - val_loss: 0.1962 - val_mean_squared_error: 0.1881\n",
      "\n",
      "Epoch 00013: val_mean_squared_error improved from 0.18980 to 0.18813, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 14/50\n",
      "3000/3000 [==============================] - 486s 162ms/step - loss: 0.1798 - mean_squared_error: 0.1798 - val_loss: 0.1995 - val_mean_squared_error: 0.2011\n",
      "\n",
      "Epoch 00014: val_mean_squared_error did not improve from 0.18813\n",
      "Epoch 15/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.1774 - mean_squared_error: 0.1774 - val_loss: 0.2695 - val_mean_squared_error: 0.2510\n",
      "\n",
      "Epoch 00015: val_mean_squared_error did not improve from 0.18813\n",
      "Epoch 16/50\n",
      "3000/3000 [==============================] - 487s 162ms/step - loss: 0.1770 - mean_squared_error: 0.1770 - val_loss: 0.1529 - val_mean_squared_error: 0.1853\n",
      "\n",
      "Epoch 00016: val_mean_squared_error improved from 0.18813 to 0.18527, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 17/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1722 - mean_squared_error: 0.1722 - val_loss: 0.2212 - val_mean_squared_error: 0.1845\n",
      "\n",
      "Epoch 00017: val_mean_squared_error improved from 0.18527 to 0.18449, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 18/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1710 - mean_squared_error: 0.1710 - val_loss: 0.1999 - val_mean_squared_error: 0.1815\n",
      "\n",
      "Epoch 00018: val_mean_squared_error improved from 0.18449 to 0.18150, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 19/50\n",
      "3000/3000 [==============================] - 489s 163ms/step - loss: 0.1698 - mean_squared_error: 0.1698 - val_loss: 0.1923 - val_mean_squared_error: 0.1992\n",
      "\n",
      "Epoch 00019: val_mean_squared_error did not improve from 0.18150\n",
      "Epoch 20/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1664 - mean_squared_error: 0.1664 - val_loss: 0.1374 - val_mean_squared_error: 0.1739\n",
      "\n",
      "Epoch 00020: val_mean_squared_error improved from 0.18150 to 0.17386, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 21/50\n",
      "3000/3000 [==============================] - 497s 166ms/step - loss: 0.1663 - mean_squared_error: 0.1663 - val_loss: 0.1754 - val_mean_squared_error: 0.1791\n",
      "\n",
      "Epoch 00021: val_mean_squared_error did not improve from 0.17386\n",
      "Epoch 22/50\n",
      "3000/3000 [==============================] - 488s 163ms/step - loss: 0.1629 - mean_squared_error: 0.1629 - val_loss: 0.1486 - val_mean_squared_error: 0.1709\n",
      "\n",
      "Epoch 00022: val_mean_squared_error improved from 0.17386 to 0.17095, saving model to 4_10_actualq1_combined_19.h5\n",
      "Epoch 23/50\n",
      "3000/3000 [==============================] - 495s 165ms/step - loss: 0.1617 - mean_squared_error: 0.1617 - val_loss: 0.2450 - val_mean_squared_error: 0.1755\n",
      "\n",
      "Epoch 00023: val_mean_squared_error did not improve from 0.17095\n",
      "Epoch 24/50\n",
      "3000/3000 [==============================] - 488s 163ms/step - loss: 0.1617 - mean_squared_error: 0.1617 - val_loss: 0.1751 - val_mean_squared_error: 0.1842\n",
      "\n",
      "Epoch 00024: val_mean_squared_error did not improve from 0.17095\n",
      "Epoch 25/50\n",
      "3000/3000 [==============================] - 496s 165ms/step - loss: 0.1587 - mean_squared_error: 0.1587 - val_loss: 0.1762 - val_mean_squared_error: 0.1745\n",
      "\n",
      "Epoch 00025: val_mean_squared_error did not improve from 0.17095\n",
      "Epoch 00025: early stopping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/keras/engine/saving.py:341: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.\n",
      "  warnings.warn('No training configuration found in save file: '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4_10_actualq1_combined_19.h5 0.11679556 0.116795585\n",
      "4_10_actualq1_model19\n"
     ]
    }
   ],
   "source": [
    "def train_model(fname='model.h5'):\n",
    "    combined_net = create_model_bidi()\n",
    "    \n",
    "    callbacks = [\n",
    "        EarlyStopping(monitor='val_mean_squared_error', mode='min', verbose=1, patience=3),\n",
    "        ModelCheckpoint(fname, monitor='val_mean_squared_error', mode='min', verbose=1, save_best_only=True)\n",
    "    ]\n",
    "\n",
    "    combined_net.fit_generator(train_gen,\n",
    "                                steps_per_epoch=3000,\n",
    "                epochs=50,\n",
    "                callbacks=callbacks,\n",
    "                validation_data=test_gen,\n",
    "                shuffle=True,\n",
    "                workers=1,\n",
    "                max_queue_size=10,\n",
    "                use_multiprocessing=False\n",
    "              )\n",
    "    return combined_net\n",
    "\n",
    "for i in range(14, 20):\n",
    "    train_model(data_folder + '_combined_' + str(i) + '.h5')\n",
    "    publish_model(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/keras/engine/saving.py:341: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.\n",
      "  warnings.warn('No training configuration found in save file: '\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4_7_actualq3_combined_10.h5 0.17735313047530865 0.1773531122255772\n",
      "4_7_actualq3_model10\n"
     ]
    }
   ],
   "source": [
    "publish_model(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
